tmt/steps/provision/__init__.py

```diff
@@ -260,8 +262,18 @@ def __init__(self, data, name=None, parent=None):
         self.role = None
         # Flag to indicate localhost guest, requires special handling
         self.localhost = False
+        # Guest is invalid by default. If it is ready, set it to be True
+        self.is_valid = False
         self.load(data)
 
+    def __bool__(self):
+        """ Detect guest is valid or not """
+        return self.is_valid
+
+    def ready(self, flag=True):
```

Comment:
My 2 cents: this is unnecessarily complex, and a trivial `is_ready` flag would do exactly the same job with much less complexity:

```python
                    if isinstance(phase, ProvisionPlugin):
                        guest = phase.guest()
                        if guest.is_ready:
                            self._guests.append(guest)

...

        if not url:
            self.is_ready = False
            raise ProvisionError(f"Could not map '{name}' to compose.")

        self.is_ready = True
        return url
```

There is no need to add a dedicated method (`ready()`) whose one and only job are setting a property to a given value - `self.is_ready = True/False` would do the trick with much less hassle -, or `__bool__()` whose only job is to *retrieve* the said property. And it'd get rid of one level of abstraction ("ready" as a facade for "is_valid").

A getter/setter approach which is common in Java is much less common in Python lands, and it becomes useful when the value needs to be computed/cached or when there are side effects involved, and even then there's a `property` decorator to help:

```python
class Guest:
  @property
  def is_ready(self) -> bool:
    # do some hard computations and find out whether the guest is ready
    return self._compute_readiness()

  @is_ready.setter
  def is_ready(self, value: bool) -> None:
    # Update readiness...
    self._update_readiness(value)
    # ... and trigger actions we need to trigger when readiness changes, e.g.:
    self._reset_connection()
    self._erase_host_hdd()
```

Unless one needs such a complex behavior when a property is read and/or written, a trivial boolean flag is a perfectly fine solution.

Reply:
@happz, thanks for your quick review. I have updated the related code by using `guest.is_ready` instead. Please help to take a look again, thanks!

Reply:
Much better, thank you!

Reply:
@happz Why new property when `__bool__` will do the job? It is implemented by default, no breakage of api will happen - just testcloud will test it properly so invalid guests are not added.

Reply:
I think there are two use cases - one needs to know if guest returned from provision is valid (== bool(guest)) and there is liveliness probe we miss but isn't in the scope for fixing #1038

Reply:
> Why new property when __bool__ will do the job? It is implemented by default, no breakage of api will happen - just testcloud will test it properly so invalid guests are not added.

I don't follow, how is `__bool__()` "implemented by default"? Reading the patch, it seemed to me `__bool__()` was a new addition, to support new `bool(guest)` coercion.

Anyway, my comments were mostly against the overcomplicated design: yes, there was `__bool__()`, therefore `bool(guest)` was possible, but this `__bool__()` was returning a newly added attribute - so what's the point in adding an extra layer of `__bool__()` when going to this attribute directly? It was added anyway, so let's give it a more fitting name and don't hide it behind `__bool__()`.

My other comment would be against `bool(guest)` in general since it's kind of obfuscating what such a "guest is true-ish" means (is it running? or just loaded? or simply in a consistent state which can be stored on disk?) - I for one find `guest.is_ready` explicit and therefore understandable. I did not raise this point in my first comment, because the original implementation was already too complicated.

Reply:
I meant that `__bool__` is defined for each python object and is already used at https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/__init__.py#L139 . The original patch just changed implementation of this method to avoid adding broken-from-the-beginning guest instances.

However you are right that bool is a bit cryptic and is_ready() makes more sense. 

BTW, shouldn't it be a method instead? It's value can be trusted only for the moment it is called and it ideally should do more than just returning instance value.

Reply:
> BTW, shouldn't it be a method instead? It's value can be trusted only for the moment it is called and it ideally should do more than just returning instance value.

As a `Guest` user, `is_ready` as a flag makes perfect sense to me - if there is a code that must be called to acquire the value, maybe reading some APIS, TBH, I don't care. It can be either `True` or `False`, and exactly *how* a class gets its answer for me is not important to me. Making it a method, i.e. `guest.is_ready()` instead of `guest.is_ready` seems like an unnecessary complication for users who need to remember they have to perform a function call instead of reading an attribute. If there's a code to run, `Guest` class and subclasses can easily perform such code in the background thanks to `@property` decorator. It would remain easier for `Guest` users while giving classes enough power to run more stuff to deduce whether the instance is ready or not. It does not have to be callable as a method though - `@property` and `guest.is_ready` is perfectly fine with me.

Ok, how about a default implementation in `Guest` class:

```python
@property
def is_ready(self) -> bool:
  return False
  # or maybe even better this, because base class has absolutely no idea when a guest becomes ready:
  raise NotImplementedError()
```

With child classes - like `GuestTestcloud` - adding their own implementation that would sometimes return `True`, like this:

```python
@property
def is_ready(self) -> bool:
  return self._instance is not None and testcloud.instance._find_domain(self._instance.name, self._instance.connection) == 'running'
```

That would be also nice. I'd vote for using `@property` to make things easier for users and open for classes. We will still have to amend all provision plugins, from `artemis.py` to `testcloud.py`.

Reply:
@happz, Since we will implement 
```python
@property
def is_ready(self) -> bool:
...
```
Do we still need the property `is_ready` OR  we should use `_is_ready` instead?

Reply:
> Making it a method, i.e. `guest.is_ready()` instead of `guest.is_ready` seems like an unnecessary complication for users who need to remember they have to perform a function call instead of reading an attribute.

I have it exactly opposite way - I prefer to call class method. (just to find out that code uses property for some reason).
However properties seem to be preferred here so I'm out of luck, lets' go with the property then.

Reply:
I see. Well, I like the boundary, the idea that from the outside, it's an attribute-like name, and I can read or write it. If there is anything to be run when reading or writing the attribute, it's not my problem, thanks to `@property` it's completely up to the class implementation to take care of such "details". If it's a trivial attribute today and becomes a `@property` next month, nothing changes for me as a user, class' API remains the same.

Reply:
Yes, boundaries are crucial and that is another thing why I prefer method - it is absolutely clear that it is read only and one doesn't care what object does internally to return its value.